### PR TITLE
[SDag][ARM][RISCV] Allow lowering CTPOP into a libcall

### PR DESCRIPTION
**Pull request by s-barannikov**
The main change is adding CTPOP to `RuntimeLibcalls.def` to allow
targets to use LibCall action for CTPOP. DAG legalizers are changed
accordingly.

### CODE REVIEW

**Comment by tschuett** [comment]
Dead code?

**Comment by arsenm** [comment]
> but I don't know if the libcall is available in their libgcc

Should be easy to find out?

**Comment by s-barannikov** [comment]
> > but I don't know if the libcall is available in their libgcc
> 
> Should be easy to find out?

Right. It appears it is available and gcc uses it for `__builtin_popcount[l,ll]` even when optimizing for speed (-Ofast).
There is no 128-bit version though.
I'll update the PR.

File: llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64xtheadbb.ll [hunk 1]
```diff
@@ -24,31 +24,14 @@ define signext i32 @ctlz_i32(i32 signext %a) nounwind {
...
-; RV64I-NEXT:    addiw a1, a1, -241
-; RV64I-NEXT:    and a0, a0, a1
-; RV64I-NEXT:    lui a1, 4112
-; RV64I-NEXT:    addiw a1, a1, 257
-; RV64I-NEXT:    call __muldi3
-; RV64I-NEXT:    srliw a0, a0, 24
+; RV64I-NEXT:    call __popcountsi2
 ; RV64I-NEXT:    ld ra, 8(sp) # 8-byte Folded Reload
 ; RV64I-NEXT:    addi sp, sp, 16
+; RV64I-NEXT:    sext.w a0, a0
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64xtheadbb.ll:30]: __popcountsi2 should be returning a sign extended result right?

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64xtheadbb.ll:30]: Nevermind, this is probably a bug specific to rv64-legal-i32 which is still experimental.

Comment (s-barannikov) [reply llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64xtheadbb.ll:30]: I tried setting `IsSExt` to true in `MakeLibCallOptions`, but that changed nothing.

**Comment by s-barannikov** [comment]
ping

File: llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp [hunk 2]
```diff
@@ -140,12 +140,9 @@ class SelectionDAGLegalize {
...
                        SmallVectorImpl<SDValue> &Results);
-  SDValue ExpandIntLibCall(SDNode *Node, bool isSigned,
-                           RTLIB::Libcall Call_I8,
-                           RTLIB::Libcall Call_I16,
-                           RTLIB::Libcall Call_I32,
-                           RTLIB::Libcall Call_I64,
-                           RTLIB::Libcall Call_I128);
+  SDValue ExpandIntLibCall(SDNode *Node, bool IsSigned, RTLIB::Libcall Call_I8,
+                           RTLIB::Libcall Call_I16, RTLIB::Libcall Call_I32,
+                           RTLIB::Libcall Call_I64, RTLIB::Libcall Call_I128);
```
Comment (jhuber6) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:145]: Looks like a lot of unrelated format changes?

Comment (s-barannikov) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:145]: I renamed isSigned -> IsSigned while I was here, clang-format insisted that I reformat this code fragment.
Do I need to revert these changes?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp [hunk 3]
```diff
@@ -3847,15 +3847,36 @@ void DAGTypeLegalizer::ExpandIntRes_CTLZ(SDNode *N,
...
 
-void DAGTypeLegalizer::ExpandIntRes_CTPOP(SDNode *N,
-                                          SDValue &Lo, SDValue &Hi) {
-  SDLoc dl(N);
+void DAGTypeLegalizer::ExpandIntRes_CTPOP(SDNode *N, SDValue &Lo, SDValue &Hi) {
+  SDValue Op = N->getOperand(0);
+  EVT VT = N->getValueType(0);
+  SDLoc DL(N);
+
+  // If the narrow CTPOP is not supported by the target, try to convert it
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:3855]: Do we really have to handle the libcall logic during type legalization? Can this just adjust the type, and leave the libcall for the libcall action

File: llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp (hunk 2 above)
Comment (jhuber6) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:145]: Ah, didn't notice the capital letter change. I don't think it's necessary to change then.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp (hunk 3 above)
Comment (s-barannikov) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:3855]: If we expand it later, we will end up with two libcalls for an illegal type instead of one (best case), or those two libcalls will be expanded into scalar code. You can see the difference in [this commit](https://github.com/llvm/llvm-project/pull/99752/commits/f90a6b46a8a9639bad8f2aa826150a47727d6b82).
This is primarily for 32-bit targets calling i64 CTPOP.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp [hunk 4]
```diff
@@ -3847,15 +3847,36 @@ void DAGTypeLegalizer::ExpandIntRes_CTLZ(SDNode *N,
...
+  // to a libcall.
+  EVT ExpandedVT = TLI.getTypeToExpandTo(*DAG.getContext(), VT);
+  if (!TLI.isOperationLegalOrCustom(ISD::CTPOP, ExpandedVT)) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:3858]: Should at least check if this is the libcall action

Comment (s-barannikov) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:3858]: I was thinking about this too, but I found [only one](https://github.com/llvm/llvm-project/blob/bb8a735facb710a07601d3bc9c45316f0de9d68d/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#L1344) place where LibCall action is checked. That made me wonder if this is the right thing to do. Hopefully someone can tell for sure.

Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:3858]: Yes. This is a stalled migration. Ideally the Expand action would never produce a libcall.

File: llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64xtheadbb.ll (hunk 1 above)
Comment (s-barannikov) [reply llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64xtheadbb.ll]: It appears 32-bit libcall is not available in 64-bit libgcc. I reverted to Expand action for rv64-legal-i32.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:145]: Can precommit this unrelated change

**Review by arsenm — APPROVED** [review]

**Comment by s-barannikov** [comment]
@topperc 
Please take a look. Are RISC-V changes OK?

**Comment by topperc** [comment]
> @topperc Please take a look. Are RISC-V changes OK?

RISC-V changes look ok.

**Comment by s-barannikov** [comment]
This is likely causing armv7/v8 stage2 build failures (other changes don't look related):
https://lab.llvm.org/buildbot/#/builders/122/builds/389
https://lab.llvm.org/buildbot/#/builders/79/builds/552

I don't have an ARM machine handy to investigate, so I'll just revert the ARM changes and see if it helps.

**Comment by DavidSpickett** [comment]
@antmox should be able to help you figure it out, he's maintaining the bots this week.

**Comment by s-barannikov** [comment]
@antmox Could you please take a look? I've already [reverted](https://github.com/llvm/llvm-project/pull/101673) the changes, but I'd like to reland them if they're not the reason of the failures.

**Comment by s-barannikov** [comment]
The reversal made the bots green.

I think I understood what the issue is. All popcount libcalls returns `int`, but ISD::CTPOP returns the type of the argument, which can be larger. In type legalizer I need to pass the correct return type to `makeLibCall` and sign-extend the result afterwards.

I'll revert the rest of the patch because the issue should affect RISC-V as well.
